[inductor] added a config to always add tensor constants

chenyang78 opened the pull request:
Summary:
In some scenarios, we want to update constants at runtime.
In such cases, we have to keep the original constants in
the generated code without applying any constant-inlining
optimizations.

This PR adds a config to force us to add tensor constants.

Differential Revision: D49895154




cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @kadeng @muchulee8 @aakhundov @ColinPeppler

jansel commented on torch/_inductor/graph.py:
```diff
@@ -528,6 +528,9 @@ def allocate(name):
                 name = f"constant_{name}"
+            # We may generate a var name for each constant in the codegen.
+            # Let's only keep sane characters.
+            name = re.sub(r"[^a-zA-Z0-9_]", "_", name)
```
Where are the non-indent chars coming from?  Could this result in name collisions?

chenyang78 replied:
The special chars come from the model. For example, some model that we are dealing with has a bias named like this ```abc:XYZ```, where `:` is not allowed in a var name in C/C++. 

Very good point regarding the name collisions! Yes, I think we could have name collisions, although the chance might be low. Perhaps we could add a counter suffix to make a unique name? e.g.

```
name = re.sub(r"[^a-zA-Z0-9_]", "_", name)
name = f"{name}_{next(self.constant_id)"
```

What do you think? Thanks.

mikekgfb replied:
This is generated code and doesn't bleed into how we should store weights for weight refresh (which is presumably the unmodified original name as a string).  Do any users need to reference the these generated names (which which case adding a suffix should be a last resort in case of duplicates, and track previously seen names), if it's just for readability during debug then adding the suffix doesn't appear to have much if any downside.

jansel replied:
Let's add a suffix and also check for collision.

chenyang78 replied:
I don't think the user needs to reference the generated names.

muchulee8 replied:
Let's use `while` for adding suffix and checking.
Something like:
```
prefix = re.sub(r"[^a-zA-Z0-9_]", "_", name)
name = prefix
cnt = 0
while name in self.constants:
  name = f"{prefix}_{cnt}"
  cnt += 1
```
In this way the name is more tracible to the original model, since it's easy to see integers appended to some variable for un-named vars (is `add_0` a constant or the result of some add operation?)
Also, I don't really need to `assert` if our artificial name collides, we could just create a new one instead.

chenyang78 replied:
Good point! Fixed. Thanks.

Commit:
[inductor] added a config to always add tensor constants (#110491)

Summary:

In some scenarios, we want to update constants at runtime.
In such cases, we have to keep the original constants in
the generated code without applying any constant-inlining
optimizations.

This PR adds a config to force us to add tensor constants.

Reviewed By: muchulee8

Differential Revision: D49895154
